### PR TITLE
probe: add probe to fill capacity for the aws ebs disk

### PR DESCRIPTION
**Pull request by sch00lb0y**
ebs probe added to fill the capacity of the disk
#83 
Signed-off-by: balaji <balaji@avinetworks.com>

### CODE REVIEW

File: cmd/probe/ebsprobe.go [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+func newEbsProbe() *ebsProbe {
+	return &ebsProbe{}
+}
+
+// Just to fullfill the interface
+func (ep *ebsProbe) Start() {}
+
+// fillDiskDetails fill the capacity of the disk
+func (ep *ebsProbe) FillDiskDetails(d *controller.DiskInfo) {
+	sysPath := d.ProbeIdentifiers.UdevIdentifier
```
Comment (shovanmaity) [inline cmd/probe/ebsprobe.go]: add one check if capacity is not present then update it else return.

File: cmd/probe/ebsprobe.go [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+
+var (
+	ebsProbeName  = "AWS EBS Probe"
+	ebsProbeState = defaultEnabled
+)
+
+var ebsProbeRegister = func() {
+	ctrl := <-controller.ControllerBroadcastChannel
+	if ctrl == nil {
+		glog.Error("unable to configure", udevProbeName)
```
Comment (shovanmaity) [inline cmd/probe/ebsprobe.go]: change probe name

File: pkg/udev/common.go [hunk 3]
```diff
@@ -58,6 +58,7 @@ const (
 	BY_ID_LINK          = "by-id"           // by-path devlink contains this string
 	BY_PATH_LINK        = "by-path"         // by-path devlink contains this string
 	LINK_ID_INDEX       = 4                 // this is used to get link index from dev link
+	SIZE                = "SIZE"            // block size
```
Comment (shovanmaity) [inline pkg/udev/common.go]: change it with `UDEV-` prefix

File: pkg/udev/common.go [hunk 4]
```diff
@@ -68,6 +69,7 @@ type UdevDiskDetails struct {
 	Path           string   // Path is Path of a disk.
 	ByIdDevLinks   []string // ByIdDevLinks contains by-id devlinks
 	ByPathDevLinks []string // ByPathDevLinks contains by-path devlinks
+	Size           string   // Block size
```
Comment (shovanmaity) [inline pkg/udev/common.go]: Size contains number of block/sector in a disk.

File: pkg/udev/common.go [hunk 5]
```diff
@@ -85,6 +87,7 @@ func (device *UdevDevice) DiskInfoFromLibudev() UdevDiskDetails {
 		Path:           device.GetPropertyValue(UDEV_DEVNAME),
 		ByIdDevLinks:   devLinks[BY_ID_LINK],
 		ByPathDevLinks: devLinks[BY_PATH_LINK],
+		Size:           device.GetSysattrValue(SIZE),
```
Comment (shovanmaity) [inline pkg/udev/common.go]: You need to update ut for this function also.

File: cmd/probe/probe.go [hunk 6]
```diff
@@ -27,7 +27,7 @@ const (
 )
 
 // RegisteredProbes contains register function of probes which we want to register
-var RegisteredProbes = []func(){smartProbeRegister, udevProbeRegister}
+var RegisteredProbes = []func(){smartProbeRegister, udevProbeRegister, ebsProbeRegister}
```
Comment (shovanmaity) [inline cmd/probe/probe.go]: break this line into multiple lines [like this](https://github.com/openebs/node-disk-manager/blob/367dcc004c55b84f7de442f9631a5f027f488fcd/cmd/filter/filter.go#L29)

**Comment by sch00lb0y** [comment]
@shovanmaity Thanks for your review. I've updated the PR. 

Let me know, If any more changes needed. :)

**Comment by kmova** [comment]
Hi @sch00lb0y !! Glad to see you back as a contributor. This PR looks good!  One thought about the naming of the probe - shall we call it - capacityprobe instead of ebs-probe. I believe this will work for non-ebs disks as well. What do you think?

**Comment by sch00lb0y** [comment]
Hi @kmova, 
Thank you. Yesterday, I totally enjoyed the HacktoberFest and met all the amazing people behind OpenEBS. It was great meeting you all. 

I've renamed the probe. Please have a look.

**Comment by sch00lb0y** [comment]
@shovanmaity updated :)

File: cmd/controller/disk.go [hunk 7]
```diff
@@ -42,6 +42,7 @@ type DiskInfo struct {
 	LogicalSectorSize uint32          // LogicalSectorSize is the Logical size of a disk in bytes
 	Compliance        string          // Compliance is implemented specifications version i.e. SPC-1, SPC-2, etc
 	DiskType          string          // DiskType represents the type of disk like Disk, Sparse etc.,
+	Size              string          // Size contains number of block/sector in a disk
```
Comment (sagarkrsd) [inline cmd/controller/disk.go]: @sch00lb0y Shouldn't the variable name should be something else than size , such as NoOfBlocks or something else ?

**Comment by sch00lb0y** [comment]
@sagarkrsd I've changed the variable name

File: cmd/probe/capacityprobe.go [hunk 8]
```diff
@@ -0,0 +1,99 @@
...
+
+// FillDiskDetails updates the capacity of the disk , if the capacity is
+// not populated.
+func (cp *capacityProbe) FillDiskDetails(d *controller.DiskInfo) {
+
+	if d.Capacity != 0 {
+		return
+	}
+	sysPath := d.ProbeIdentifiers.UdevIdentifier
+	blockSize, err := strconv.ParseInt(d.NoOfBlocks, 10, 64)
```
Comment (shovanmaity) [inline cmd/probe/capacityprobe.go]: looks like we are not getting no of block using udev probe so you need to fetch both block count and block size from file.
```
shovan@shovan-Latitude-3560:/sys/dev/block/8:0$ cat size 
976773168
shovan@shovan-Latitude-3560:/sys/dev/block/8:0$ cat queue/hw_sector_size 
512
```
update tthis pr according this.

File: cmd/probe/capacityprobe.go [hunk 9]
```diff
@@ -0,0 +1,99 @@
...
+	if err != nil {
+		glog.Error("unable to parse the block size ", err)
+		return
+	}
+	b, err := ioutil.ReadFile(sysPath + "/queue/hw_sector_size")
+	if err != nil {
+		glog.Error("unable to read sector size", err)
+		return
+	}
+	sectorSize, err := strconv.ParseInt(string(b), 10, 64)
```
Comment (shovanmaity) [inline cmd/probe/capacityprobe.go]: You need to trim `\n` else it will give error

File: cmd/probe/capacityprobe.go [hunk 10]
```diff
@@ -0,0 +1,99 @@
...
+	if err != nil {
+		glog.Error("unable to parse the sector size", err)
+		return
+	}
+	d.Capacity = uint64(blockSize * sectorSize)
```
Comment (shovanmaity) [inline cmd/probe/capacityprobe.go]: check if logical sector size is not set then set it with `sectorSize`

File: pkg/udev/common.go [hunk 11]
```diff
@@ -85,6 +87,7 @@ func (device *UdevDevice) DiskInfoFromLibudev() UdevDiskDetails {
 		Path:           device.GetPropertyValue(UDEV_DEVNAME),
 		ByIdDevLinks:   devLinks[BY_ID_LINK],
 		ByPathDevLinks: devLinks[BY_PATH_LINK],
+		NoOfBlocks:     device.GetSysattrValue(UDEV_SIZE),
```
Comment (shovanmaity) [inline pkg/udev/common.go]: remove this once you update your logic

File: pkg/udev/common.go [hunk 12]
```diff
@@ -68,6 +69,7 @@ type UdevDiskDetails struct {
 	Path           string   // Path is Path of a disk.
 	ByIdDevLinks   []string // ByIdDevLinks contains by-id devlinks
 	ByPathDevLinks []string // ByPathDevLinks contains by-path devlinks
+	NoOfBlocks     string   // NoOfBlocks contains number of block/sector in a disk.
```
Comment (shovanmaity) [inline pkg/udev/common.go]: remove this once you update your logic

File: pkg/udev/common.go [hunk 13]
```diff
@@ -58,6 +58,7 @@ const (
 	BY_ID_LINK          = "by-id"           // by-path devlink contains this string
 	BY_PATH_LINK        = "by-path"         // by-path devlink contains this string
 	LINK_ID_INDEX       = 4                 // this is used to get link index from dev link
+	UDEV_SIZE           = "SIZE"            // udev attribute to get number of block/sector in disk
```
Comment (shovanmaity) [inline pkg/udev/common.go]: remove this once you update your logic

File: cmd/probe/udevprobe.go [hunk 14]
```diff
@@ -170,6 +170,7 @@ func (up *udevProbe) FillDiskDetails(d *controller.DiskInfo) {
 	d.Vendor = udevDiskDetails.Vendor
 	d.ByIdDevLinks = udevDiskDetails.ByIdDevLinks
 	d.ByPathDevLinks = udevDiskDetails.ByPathDevLinks
+	d.NoOfBlocks = udevDiskDetails.NoOfBlocks
```
Comment (shovanmaity) [inline cmd/probe/udevprobe.go]: remove this once you update your logic

File: cmd/probe/capacityprobe.go (hunk 10 above)
Comment (sch00lb0y) [reply cmd/probe/capacityprobe.go]: @shovanmaity could you explain this?

Comment (shovanmaity) [reply cmd/probe/capacityprobe.go]: We are not getting logical sector size of disk also using `udev` and `smart` probe for some cases. This probe is going to fetch that by reading system file to get capacity of disk, it should update `logical sector size` also. It is not a part of issue #83 but I will suggest you to update that value if it is not set. You need to check if [logical sector size](https://github.com/openebs/node-disk-manager/blob/4655867f989bed4a56c9c22966ac8907c960de22/cmd/controller/disk.go#L42) attribute is not set then set it using this probe. (same as `d.Capacity`)

**Comment by sch00lb0y** [comment]
@shovanmaity I've updated, but I've not tested in k8s :(. 
Is it okay, if I validate this in my local minikube? or should I've to spin a aws cluster.

**Comment by shovanmaity** [comment]
> @shovanmaity I've updated, but I've not tested in k8s :(.
> Is it okay, if I validate this in my local minikube? or should I've to spin a aws cluster.

@sch00lb0y we have tested it in aws cluster. :tada:

**Comment by princerachit** [comment]
Hi @sch00lb0y, Thank you very much for participating in Hacktoberfest and contributing to OpenEBS and in general to the Open Source. If you have not already starred/liked the following repositories,  you may want to do so now :slightly_smiling_face:

[OpenEBS](https://github.com/openebs/openebs/stargazers) [Maya](https://github.com/openebs/maya/stargazers) [Litmus](https://github.com/openebs/litmus/stargazers) [NDM](https://github.com/openebs/node-disk-manager/stargazers) [WeaveScope](https://github.com/weaveworks/scope/stargazers) [LongHorn](https://github.com/openebs/longhorn/stargazers) [HeptioArk](https://github.com/heptio/ark/stargazers)

**Pull request merged**